python/lsst/pipe/base/pipeline_graph/visualization/_formatting.py

```diff
@@ -0,0 +1,175 @@
+        ``get_symbol`` callback.
+
+    Returns
+    -------
+    symbol : `str`
+        A single-character symbol.
+    """
+    match node:
+        case NodeKey(node_type=NodeType.TASK):
+            return "■"
+        case NodeKey(node_type=NodeType.DATASET_TYPE):
+            return "○"
+        case NodeKey(node_type=NodeType.TASK_INIT):
+            return "▣"
+        case MergedNodeKey(node_type=NodeType.TASK):
+            return "▥"
+        case MergedNodeKey(node_type=NodeType.DATASET_TYPE):
+            return "◍"
+        case MergedNodeKey(node_type=NodeType.TASK_INIT):
+            return "▤"
+    raise ValueError(f"Unexpected node key: {node} of type {type(node)}.")
+
+
+class GetNodeText:
+    """A callback for the `Printer` class's `get_text` callback that
+    prints detailed information about a node and can defer long entries to
+    a footnote.
+
+    Parameters
+    ----------
+    xgraph : `networkx.DiGraph` or `networkx.MultiDiGraph`
+        NetworkX export of a `.PipelineGraph` that is being displayed.
+    options : `NodeAttributeOptions`
+        Options for how much information to display.
+    width : `int` or `None`
+        Number of display columns that node text can occupy.  `None` for
+        unlimited.
+    """
+
+    def __init__(
+        self,
+        xgraph: networkx.DiGraph | networkx.MultiDiGraph,
+        options: NodeAttributeOptions,
+        width: int | None,
+    ):
+        self.xgraph = xgraph
+        self.options = options
+        self.width = width
+        self.deferred: list[tuple[str, tuple[str, str], list[str]]] = []
+
+    def __call__(self, node: DisplayNodeKey, x: int, style: tuple[str, str]) -> str:
+        state = self.xgraph.nodes[node]
+        terms: list[str] = [f"{node}:" if self.options else str(node)]
+        if self.options.dimensions and node.node_type != NodeType.TASK_INIT:
+            terms.append(self.format_dimensions(state["dimensions"]))
+        if (
+            self.options.task_classes
+            and node.node_type is NodeType.TASK
+            or node.node_type is NodeType.TASK_INIT
+        ):
+            terms.append(self.format_task_class(state["task_class_name"]))
+        if self.options.storage_classes and node.node_type is NodeType.DATASET_TYPE:
+            terms.append(state["storage_class_name"])
+        description = " ".join(terms)
+        if self.width and len(description) > self.width:
+            index = f"[{len(self.deferred) + 1}]"
+            self.deferred.append((index, style, terms))
+            return f"{description[:self.width - len(index) - 6]}...{style[0]}{index}{style[1]} "
+        return description
+
+    def format_dimensions(self, dimensions: DimensionGraph) -> str:
+        """Format the dimensions of a task or dataset type node."""
+        match self.options.dimensions:
+            case "full":
+                return str(dimensions)
+            case "concise":
```

Comment:
Here's the second half of the answer to the question about which dimensions are shown: the algorithm that removes unnecessary dimensions in the "concise" version.